### PR TITLE
add prediction script

### PR DESCRIPTION
**Pull request by wenmengzhou**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

# Motivation
add tools/predict.py for local single worker and multi-worker prediction with following features:
1.  support odps table read and write
2.  high performance using multi-threading for each process such as image downloading, image decoding, prediction.
3. support oss read and write using pai-dlc 

# Modification
Please briefly describe what modification is made in this PR.

# Backward-Compatibility-Breaking (Optional)
Does the modification introduce changes that break the backward-compatibility? If so, please describe how it breaks the compatibility in this PR.

# Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

# Pull-Request Todo List
1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: tools/predict.py [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+                          TableReadProcess, TableWriteProcess)
+from mmcv.runner import init_dist
+
+from easycv.utils.dist_utils import get_dist_info
+from easycv.utils.logger import get_root_logger
+
+try:
+    import easy_predict
+except ModuleNotFoundError:
+    print('please install easy_predict first')
```
Comment (Cathy0908) [inline tools/predict.py:25]: add easy_predict whl to requirements

File: tools/predict.py [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+                ofile.write(l)
+
+
+def build_and_run_file_io(args):
+    # distribute info
+    rank, world_size = get_dist_info()
+    worker_id = rank
+
+    input_oss_file_new_host = args.input_file + '.tmp%d' % worker_id
+    replace_oss_with_local_path(args.input_file, input_oss_file_new_host,
```
Comment (Cathy0908) [inline tools/predict.py:260]: 这个是为了适配什么情况？

File: tools/predict.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    num_worker = world_size
+    print(f'worker num {num_worker}')
+    print(f'worker_id {worker_id}')
+    thread_num = 1
+    batch_size = args.batch_size
+    # os.environ['CUDA_VISIBLE_DEVICES'] = str(rank % torch.cuda.device_count())
+    print(f'Local rank {args.local_rank}')
+    torch.cuda.set_device(args.local_rank)
+    predictor = init_predictor(args)
+    predict_fn = predictor.__call__ if hasattr(
```
Comment (Cathy0908) [inline tools/predict.py:272]: __call__方法很多类都有默认方法，这样用不太安全，改成如下？
```python
predict_fn = predictor.predict if callable(getattr(predictor, predict)) else predictor
```
同 build_and_run_table_io

File: tools/predict.py (hunk 2 above)
Comment (wenmengzhou) [reply tools/predict.py:260]: 输入文件里是oss路径，需要替换为挂载后的路径

File: tools/predict.py (hunk 1 above)
Comment (wenmengzhou) [reply tools/predict.py:25]: 这里不希望作为默认依赖都装上，希望用户在用的时候可以自行安装，因为不是必要的

Comment (Cathy0908) [reply tools/predict.py:25]: 那用户应该怎么安装？直接 pip install easy_predict？

File: tools/predict.py (hunk 2 above)
Comment (Cathy0908) [reply tools/predict.py:260]: 挂载的动作是云服务做的，那这个脚本只适用于云服务产品吗，本地可以运行吗？
或者用户自己在dsw或者dlc启动这个脚本，传入oss路径，没有进行挂载可以使用吗？
这个脚本适用于什么场景？

Comment (wenmengzhou) [reply tools/predict.py:260]: 本地使用本地文件路径
云上挂载可以使用oss路径

本地传入oss路径的方式没有支持

File: tools/predict.py (hunk 1 above)
Comment (wenmengzhou) [reply tools/predict.py:25]: 增加离线推理doc，里面说明easy_predict版本

**Review by Cathy0908 — APPROVED** [review]

File: tools/predict.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+                          ProcessExecutor, ResultGatherProcess,
+                          TableReadProcess, TableWriteProcess)
+from mmcv.runner import init_dist
+
+from easycv.utils.dist_utils import get_dist_info
+from easycv.utils.logger import get_root_logger
+
+try:
+    import easy_predict
+except ModuleNotFoundError:
```
Comment (wuziheng) [inline tools/predict.py:24]: 这个try catch 不放在上面的import 之前是不是会先挂啊

File: tools/predict.py (hunk 1 above)
Comment (wuziheng) [reply tools/predict.py:25]: 直接把whl链接和install命令挂载这里就好了可以么

File: docs/source/tutorials/predict.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+多机多卡运行
+
+这里多机多卡启动方式复用pytorch DDP方式
+```shell
+PYTHONPATH=. python -m torch.distributed.launch --nproc_per_node=2 --master_port=29527 \
+                    tools/predict.py \
+                    --input_file predict/test.list \
+                    --output_file predict/output.txt \
+                    --model_type YoloXPredictor \
+                    --model_path  predict/test_data/yolox/epoch_300.pt\
```
[the remark this answers is not in the document]
Comment (wenmengzhou) [reply docs/source/tutorials/predict.md:106]: 这个主要的需求场景是？

File: tools/predict.py [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+def build_and_run_table_io(args):
+    os.environ['ODPS_CONFIG_FILE_PATH'] = args.odps_config
+
+    rank, world_size = get_dist_info()
+    worker_id = rank
+    num_worker = world_size
+    print(f'worker num {num_worker}')
+    print(f'worker_id {worker_id}')
+
+    thread_num = 1
```
Comment (wuziheng) [inline tools/predict.py:337]: 多余的

File: tools/predict.py (hunk 4 above)
Comment (wenmengzhou) [reply tools/predict.py:24]: isort的自动修改导致，已fix

File: tools/predict.py (hunk 1 above)
Comment (wenmengzhou) [reply tools/predict.py:25]: done

File: tools/predict.py (hunk 6 above)
Comment (wenmengzhou) [reply tools/predict.py:337]: done

File: tools/predict.py [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+    for handler in logger.root.handlers:
+        if type(handler) is logging.StreamHandler:
+            handler.setLevel(logging.INFO)
+
+
+if __name__ == '__main__':
+    args = define_args()
+    patch_logging()
+    if args.launcher:
+        init_dist(args.launcher, backend='nccl')
```
[the remark this answers is not in the document]
Comment (wenmengzhou) [reply tools/predict.py:446]: 在参数定义和帮助文档中均说明

**Pull request merged**